geo/src/algorithm/is_convex.rs

```diff
@@ -119,14 +120,216 @@ impl<T: GeoNum> IsConvex for LineString<T> {
         if !self.is_closed() || self.0.is_empty() {
             None
         } else {
-            is_convex_shaped(&self.0[1..], allow_collinear, specific_orientation)
+            // Use sign flip algorithm which detects both non-convexity and self-intersections
+            if is_convex_sign_flips(&self.0) {
+                // If sign flips indicate convex, determine orientation
+                is_convex_shaped(&self.0[1..], allow_collinear, specific_orientation)
+            } else {
+                None
+            }
         }
     }
 
     fn is_collinear(&self) -> bool {
         self.0.is_empty()
             || is_convex_shaped(&self.0[1..], true, Some(Orientation::Collinear)).is_some()
     }
+
+    /// Test if the shape is convex.
+    fn is_convex(&self) -> bool {
+        // Convexity only makes sense for closed shapes
```

Comment:
It might make sense to put this check in `is_convex_sign_flips`. 

It's a cheap check and we don't allow "convexity" without it, right?

Reply:
Good point. `is_convex_sign_flips` accepts a slice: `[Coord<T>]`, but we can check the equality of the first and last slice elements.